### PR TITLE
Green Light Recognition

### PR DESCRIPTION
**Pull request by RobynCastro**

### CODE REVIEW

**Comment by RobynCastro** [comment]
Seems like my tests are failing because all my tests are graphical tests not unit tests (That's something I still need to do).

**Comment by garethellis0** [comment]
Hmm, I think we managed to get the wierd image path stuff sorted out over on the IGVC repo, maybe take a look at what we did with your vision testing stuff over there?

File: src/green_recognition/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,84 @@
+cmake_minimum_required(VERSION 2.8.3)
+project(green_recognition)
+
```
Comment (garethellis0) [inline src/green_recognition/CMakeLists.txt]: Just a note: I'm OK with keeping this and line detection separate for now, but we should likely eventually just put them in the same project.

File: src/green_recognition/include/CircleDetection.h [hunk 2]
```diff
@@ -0,0 +1,108 @@
+/*
+ * Created By: Robyn Castro
+ * Created On: June 15, 2017
+ * Description: Determines whether or not green is seen on
+ *              the screen then publishes a recommended twist
+ *              message.
+ * Usage:
+ * Subscribes to: camera // TODO: Determine topic names
+ * Publishes to:
```
Comment (garethellis0) [inline src/green_recognition/include/CircleDetection.h]: I'm OK if we just leave the `Subscribes to` and `Publishes to` sections out entirely. They'll probably be hard to maintain and likely get out of sync with the actual code fairly quickly.

File: src/green_recognition/include/CircleDetection.h [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+#include <ros/ros.h>
+#include <ros/ros.h>
+#include <geometry_msgs/Twist.h>
+
+// STD
+#include <vector>
+#include <string>
+
+// Snowbots
+#include <sb_utils.h>
```
Comment (garethellis0) [inline src/green_recognition/include/CircleDetection.h]: I like the organization of the `#include`'s!  :+1:

File: src/green_recognition/include/CircleDetection.h [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+using namespace cv;
+
+class GreenRecognition {
```
Comment (garethellis0) [inline src/green_recognition/include/CircleDetection.h]: Is `GreenRecognition` descriptive of what this node is? Something like `GreenLightDetection` might be a bit better. Or better yet, just `TrafficLightDetection`, as we may want to generalise this to recognise lights of any color.

Comment (garethellis0) [reply src/green_recognition/include/CircleDetection.h]: The filename should match the class name.

File: src/green_recognition/src/CircleDetection.cpp [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+    vector<int> radii;
+
+    // Convert grayscale image to black and white image
+    cv::Mat bwImage;
+    try {
+        cv::cvtColor(filtered_image, bwImage, CV_RGB2GRAY);
+    } catch(cv::Exception) {
+        bwImage = filtered_image;
+    }
+    cv::findContours(bwImage.clone(), contours, hierarchy, CV_RETR_TREE, CV_CHAIN_APPROX_NONE);
```
Comment (garethellis0) [inline src/green_recognition/src/CircleDetection.cpp]: A quick comment on what we're actually searching for in the image here would be helpful

File: src/green_recognition/include/CircleDetection.h [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+    int countCircles(const Mat &filtered_image);
+
+private:
+
+    /**
+     * Callback for the filtered image
+     *
+     * @param address of filtered image matrix
+     */
+    void subscriberCallBack(const sensor_msgs::Image::ConstPtr &image);
```
Comment (garethellis0) [inline src/green_recognition/include/CircleDetection.h]: `subscriberCallBack` is a little vague. Perhaps something like `filteredImageCallBack` might be better?

File: src/green_recognition/include/CircleDetection.h [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+     * @param image to be parsed
+     */
+    int countCircles(const Mat &filtered_image);
+
+private:
+
+    /**
+     * Callback for the filtered image
+     *
+     * @param address of filtered image matrix
```
Comment (garethellis0) [inline src/green_recognition/include/CircleDetection.h]: We can probably just simplify this to `filtered image`. It's pretty clear it's a pointer, and all images are really just some form of matrix anyhow

File: src/green_recognition/include/GreenFilter.h [hunk 8]
```diff
@@ -0,0 +1,137 @@
+/*
+ * Created By: Robyn Castro
+ * Created On: June 15, 2017
+ * Description: Anything green is mapped to white, everything
+ *              else is mapped to black.
+ *
+ * Usage:
+ * Subscribes to: camera // TODO: Determine topic names
+ * Publishes to:
```
Comment (garethellis0) [inline src/green_recognition/include/GreenFilter.h]: See above comment, I'm OK if we leave these sections out.

File: src/green_recognition/include/CircleDetection.h [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+                                   std::vector<int> radii);
+
+    /**
+     * Determines whether path contains an image.
+     *
+     */
+    void check_if_image_exist(const cv::Mat &img, const std::string &path);
+
+    /**
+     * Subscribes to the filtered camera image node
```
Comment (garethellis0) [inline src/green_recognition/include/CircleDetection.h]: nit-picky thing, but we're subscribing to a *topic*, not a *node*

File: src/green_recognition/include/CircleDetection.h [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+     */
+    image_transport::Subscriber image_sub;
+
+    /**
+     * Publishes the recommended twist message
```
Comment (garethellis0) [inline src/green_recognition/include/CircleDetection.h]: I think this is a leftover from an old node, this publisher is publishing a boolean right?

File: src/green_recognition/include/CircleDetection.h [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+     */
+    ros::Publisher boolean_pub;
```
Comment (garethellis0) [inline src/green_recognition/include/CircleDetection.h]: We may want to rename this to indicate what the boolean actually means, something like `green_light_detected_pub`

File: src/green_recognition/include/GreenFilter.h [hunk 12]
```diff
@@ -0,0 +1,137 @@
+/*
+ * Created By: Robyn Castro
+ * Created On: June 15, 2017
+ * Description: Anything green is mapped to white, everything
+ *              else is mapped to black.
+ *
```
Comment (garethellis0) [inline src/green_recognition/include/GreenFilter.h]: LOL. Ok, so I made a bunch of comments on `CircleDetection.h` above, but I think that you may be using this file now, so just apply my comments from `CircleDetection.h` where they're appropriate.

**Review by garethellis0 — DISMISSED** [review]
OK, I got partway through this review, but I think there are some old files in here, there seems to be several really similar classes under different filenames. Could either go through and remove and files that shouldn't be here, or confirm that all the files that are here should be here before I finish my review? Just tag me when you want me to take another pass

**Comment by garethellis0** [comment]
@ValRat @RobynCastro  - Is this ready for final review?

**Comment by ValRat** [comment]
Nay

**Review by ValRat — APPROVED** [review]
:vertical_traffic_light: :video_camera: :robot:

**Pull request merged**